lib/crypto/src/field/fp.rs

```diff
@@ -144,91 +305,203 @@ declare_fp!(Fp704, LIMBS_704, 704);
+    #[inline(always)]
+    #[must_use]
+    pub fn is_geq_modulus(&self) -> bool {
+        self.montgomery_form >= P::MODULUS
+    }
+
+    #[inline(always)]
+    fn subtract_modulus(&mut self) {
+        if self.is_geq_modulus() {
+            self.montgomery_form.checked_sub_assign(&Self::MODULUS);
+        }
+    }
+
+    /// Construct a new field element from its underlying
+    /// [`struct@Uint`] data type.
     #[inline]
     #[must_use]
-    pub const fn new_unchecked(element: Uint<LIMBS>) -> Self {
-        Fp {
-            residue: Residue::<ResidueParam<P, LIMBS>, LIMBS>::from_montgomery(
-                element,
-            ),
+    pub const fn new(element: Uint<N>) -> Self {
+        let mut r = Self { montgomery_form: element, phantom: PhantomData };
+        if r.ct_is_zero() {
+            r
+        } else {
+            r = r.ct_mul(&Fp { montgomery_form: P::R2, phantom: PhantomData });
+            r
         }
     }
-}
 
-impl<P: FpParams<LIMBS>, const LIMBS: usize> Hash for Fp<P, LIMBS> {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        self.residue.as_montgomery().hash(state);
+    /// Multiply `self` to `rhs` and return the result (constant).
+    const fn ct_mul(&self, rhs: &Self) -> Self {
+        let (carry, res) = self.ct_mul_without_cond_subtract(rhs);
+        if P::HAS_MODULUS_SPARE_BIT {
+            res.ct_subtract_modulus()
+        } else {
+            res.ct_carrying_sub_modulus(carry)
+        }
+    }
+
+    //// Returns true if this number is zero (constant).
+    const fn ct_is_zero(&self) -> bool {
+        self.montgomery_form.ct_is_zero()
+    }
+
+    /// Subtract modulus from `self` with carry.
+    #[inline(always)]
+    fn carrying_sub_modulus(&mut self, carry: bool) {
+        if carry || self.is_geq_modulus() {
+            self.montgomery_form.checked_sub_assign(&Self::MODULUS);
+        }
+    }
+
+    #[inline(always)]
+    const fn ct_mul_without_cond_subtract(&self, other: &Self) -> (bool, Self) {
+        let (lo, hi) =
+            self.montgomery_form.ct_widening_mul(&other.montgomery_form);
+
+        let (carry, res) = Self::widening_montgomery_reduction(lo, hi);
+        (carry, Self::new_unchecked(res))
+    }
+
+    /// Apply the Montgomery reduction to composite number represented as `lo`
+    /// and `hi`.
+    /// Returns `carry` and the result of the reduction.
+    ///
+    /// Algorithm 14.32 in Handbook of Applied Cryptography [reference].
+    ///
+    /// [reference]: https://cacr.uwaterloo.ca/hac/about/chap14.pdf
+    #[inline(always)]
+    const fn widening_montgomery_reduction(
+        mut lo: Uint<N>,
+        mut hi: Uint<N>,
+    ) -> (bool, Uint<N>) {
+        let mut carry2 = 0;
+        ct_for_unroll6!((i in 0..N) {
+            let tmp = lo.limbs[i].wrapping_mul(P::INV);
+
+            let (_, mut carry) = arithmetic::limb::mac(lo.limbs[i], tmp, P::MODULUS.limbs[0]);
+
+            ct_for_unroll6!((j in 1..N) {
+                let k = i + j;
```

Comment:
Nit: I do not like `k`, `i`, `l` for loop iterations, prefer more self-descriptive names to not mess them.